python/cudf/cudf/core/index.py

```diff
@@ -2744,6 +2748,212 @@ def categories(self):
+        assert end is not None and start is not None
+        end = end + 1
+        periods_array = cupy.asarray(cupy.arange(start, end))
+        _, bin_edges = cupy.histogram(periods_array, periods)
+        # cupy.histogram turns all arrays into a float array
+        # this can cause the dtype to be a float instead of an int
+        # the below adjusts for this
+        if cupy.all(cupy.mod(bin_edges, 1) == 0):
+            bin_edges = bin_edges.astype(int)
+        left_col = bin_edges[:-1]
+        right_col = bin_edges[1:]
+    elif freq and periods:
+        if end:
+            start = end - (freq * periods)
+        if start:
+            end = freq * periods + start
+        left_col = cupy.arange(start, end, freq)
+        assert end is not None and start is not None
+        end = end + 1
+        start = start + freq
+        right_col = cupy.arange(start, end, freq)
+    elif freq and not periods:
+        assert end is not None and start is not None
+        end = end - freq + 1
+        left_col = cupy.arange(start, end, freq)
+        end = end + freq + 1
+        start = start + freq
+        right_col = cupy.arange(start, end, freq)
+    elif start is not None and end is not None:
+        left_col = cupy.arange(start, end, freq)
+        start = start + 1
+        end = end + 1
+        right_col = cupy.arange(start, end, freq)
+    else:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, atleast two must be specified"
+        )
+    if len(right_col) == 0 or len(left_col) == 0:
+        return cudf.IntervalIndex([], closed=closed)
+
+    interval_col = column.build_interval_column(
+        left_col, right_col, closed=closed
+    )
+    return IntervalIndex(interval_col)
+
+
+class IntervalIndex(GenericIndex):
+    """
+    Immutable index of intervals that are closed on the same side.
+
+    Parameters
+    data : array-like (1-dimensional)
+    Array-like containing Interval objects from which to build the
+    IntervalIndex.
+
+    closed : {"left", "right", "both", "neither"}, default "right"
+    Whether the intervals are closed on the left-side, right-side,
+    both or neither.
+
+    dtype : dtype or None, default None
+    If None, dtype will be inferred.
+
+    copy : bool, default False
+    Copy the input data.
+
+    name : object, optional
+    Name to be stored in the index.
+
+    Returns
+    IntervalIndex
+    """
+
+    def __new__(
+        cls, data=None, closed=None, dtype=None, copy=False, name=None,
+    ) -> "IntervalIndex":
+        if copy:
+            data = column.as_column(data, dtype=dtype).copy(deep=True)
+        out = Frame.__new__(cls)
+        kwargs = _setdefault_name(data, name=name)
+        if isinstance(data, IntervalColumn):
+            data = data
+        elif isinstance(data, pd.Series) and (is_interval_dtype(data.dtype)):
+            data = column.as_column(pa.Array.from_pandas(data), dtype=dtype,)
```

Comment:
does `column.as_column` not handle a `pd.Series` of interval dtype properly? Should it?

Reply:
Yes, I was struggling to get this to work in any other way :/ I'll try see if something else could work, but yes I was running into a wall with other alternatives here.

Reply:
Whoops, actually, I tried running this with `as_column`, directly passing in the series and it works 😄 I realise when I was just letting it run through to the last else block I was getting an error. I've changed the code so that it doesn't change it to a pyarrow array!